File: MontyHall.java
import java.util.Random;

public class MontyHall {

    public static float probabilityPercent(float a, float b) {
        return (a / b) * 100.f;
    }

    public static void main(String[] args) {

        int ITERATIONS = 10_000_000;

        Random rnd = new Random();

        int successesStay = 0;
        int successesSwitch = 0;
        
        long startTime = System.nanoTime();

        for (int i = 0; i < ITERATIONS; i++) {
            boolean[] doors = {false, false, false};

            int doorIndex = rnd.nextInt(3);
            doors[doorIndex] = true;

            // Pick a door
            int choiceContenstant = rnd.nextInt(3);
        
            // Eliminate one (dealers choice)
            int choiceHost = -1;

            if (doorIndex == choiceContenstant) {
                int[] possibleChoices = {0, 0};
                int index = 0;
                for (int j = 0; j <= 2; j++) {
                    if (j != doorIndex) {
                        possibleChoices[index++] = j;
                    }

                    choiceHost = possibleChoices[rnd.nextInt(2)];
                }
            } else {
                choiceHost = 3 - (choiceContenstant + doorIndex);
            }

            // Stay
            if (doors[choiceContenstant])
                successesStay++;

            // Switch
            if (doors[3 - (choiceContenstant + choiceHost)])
                successesSwitch++;
        }

        long stopTime = System.nanoTime();
        
        float probabilityStay = probabilityPercent(successesStay, ITERATIONS);
        float probabilitySwitch = probabilityPercent(successesSwitch, ITERATIONS);

        System.out.println("Probability of success when staying: " + probabilityStay + "%");
        System.out.println("Probability of success when switching: " + probabilitySwitch + "%\n");

        System.out.println(ITERATIONS + " Iterations completed in " + (float)(stopTime - startTime)/1_000_000 + " milliseconds.\n");
    }
}

File: MontyHall.py
import random
import time

DEBUG = False

ITERATIONS = 10_000_000


def probabilityPercent(a, b):
    return (a / b) * 100


successesStay = 0
successesSwitch = 0

startTime = time.time()

for _ in range(ITERATIONS):
    doors = [False]*3

    doors[random.randint(0,2)] = True

    if DEBUG:
        print(f"Door Truth: {doors}")

    # Pick a door
    choiceContenstant = random.randint(0,2)
    if DEBUG:
        print(f"Contestant chose door #{choiceContenstant + 1}")

    if doors[(choiceContenstant + 1) % 3]:
        choiceHost = (choiceContenstant - 1) % 3
    else:
        choiceHost = (choiceContenstant + 1) % 3

    if DEBUG:
        print(f"Host revealed door #{choiceHost + 1}")

    # Stay
    if doors[choiceContenstant]:
        successesStay += 1
        if DEBUG:
            print("Winning strategy: STAY")

    if doors[3 - (choiceContenstant + choiceHost)]:
        successesSwitch += 1
        if DEBUG:
            print("Winning strategy: SWITCH")
    if DEBUG:
        print("---------------\n")

stopTime = time.time()

probabilityStay = probabilityPercent(successesStay, ITERATIONS)
probabilitySwitch = probabilityPercent(successesSwitch, ITERATIONS)

print(f"Probability of success when staying: {probabilityStay}%")
print(f"Probability of success when switching: {probabilitySwitch}%")

print(f"\n{ITERATIONS} iterations completed in {(stopTime-startTime)*1000} milliseconds.\n")
